File: backend/app/api/routes/clothing_items.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional
from app.db.database import get_db
from app.crud import clothing_items as clothing_crud
from app.schemas.clothing_items import ClothingItemsListResponse, ClothingItemResponse
import math

router = APIRouter(prefix="/api/clothing", tags=["clothing"])

@router.get("/", response_model=ClothingItemsListResponse)
async def get_clothing_items(
    page: int = Query(1, ge=1, description="페이지 번호"),
    size: int = Query(20, ge=1, le=100, description="페이지 크기"),
    sort_by: str = Query("likes", description="정렬 기준 (likes, product_id, product_name)"),
    order: str = Query("desc", description="정렬 순서 (asc, desc)"),
    category: Optional[str] = Query(None, description="카테고리 필터"),
    gender: Optional[str] = Query(None, description="성별 필터"),
    db: Session = Depends(get_db)
):
    """의류 아이템 목록 조회"""
    try:
        skip = (page - 1) * size
        items, total = clothing_crud.get_clothing_items(
            db=db, 
            skip=skip, 
            limit=size,
            sort_by=sort_by,
            order=order,
            category=category,
            gender=gender
        )
        
        total_pages = math.ceil(total / size)
        
        return ClothingItemsListResponse(
            items=items,
            total=total,
            page=page,
            size=size,
            total_pages=total_pages
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"데이터 조회 중 오류가 발생했습니다: {str(e)}")
    
@router.get("/categories", response_model=dict)
async def get_categories(db: Session = Depends(get_db)):
    """카테고리 목록 조회"""
    try:
        categories = clothing_crud.get_categories(db=db)
        return categories
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"카테고리 조회 중 오류가 발생했습니다: {str(e)}")
    
@router.get("/browse", response_model=ClothingItemsListResponse)
async def browse_clothing_items(
    page: int = Query(1, ge=1, description="페이지 번호"),
    size: int = Query(20, ge=1, le=100, description="페이지 크기"),
    sort_by: str = Query("likes", description="정렬 기준 (likes, latest, name)"),
    order: str = Query("desc", description="정렬 순서 (asc, desc)"),
    main_category: Optional[str] = Query(None, description="메인 카테고리"),
    sub_category: Optional[str] = Query(None, description="서브 카테고리"),
    gender: Optional[str] = Query(None, description="성별"),
    brand: Optional[str] = Query(None, description="브랜드"),
    search: Optional[str] = Query(None, description="검색어"),
    db: Session = Depends(get_db)
):
    """필터링과 검색이 가능한 의류 아이템 브라우징"""
    try:
        skip = (page - 1) * size
        items, total = clothing_crud.get_clothing_items_with_filters(
            db=db,
            skip=skip,
            limit=size,
            sort_by=sort_by,
            order=order,
            main_category=main_category,
            sub_category=sub_category,
            gender=gender,
            brand=brand,
            search=search
        )
        
        total_pages = math.ceil(total / size)
        
        return ClothingItemsListResponse(
            items=items,
            total=total,
            page=page,
            size=size,
            total_pages=total_pages
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"상품 브라우징 중 오류가 발생했습니다: {str(e)}")

@router.get("/popular", response_model=list[ClothingItemResponse])
async def get_popular_items(
    limit: int = Query(6, ge=1, le=20, description="조회할 상품 수"),
    db: Session = Depends(get_db)
):
    """인기 상품 조회"""
    try:
        items = clothing_crud.get_popular_items(db=db, limit=limit)
        return items
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"인기 상품 조회 중 오류가 발생했습니다: {str(e)}")

@router.get("/latest", response_model=list[ClothingItemResponse])
async def get_latest_items(
    limit: int = Query(6, ge=1, le=20, description="조회할 상품 수"),
    db: Session = Depends(get_db)
):
    """최신 상품 조회"""
    try:
        items = clothing_crud.get_latest_items(db=db, limit=limit)
        return items
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"최신 상품 조회 중 오류가 발생했습니다: {str(e)}")

@router.get("/search", response_model=ClothingItemsListResponse)
async def search_clothing_items(
    q: str = Query(..., description="검색어"),
    page: int = Query(1, ge=1, description="페이지 번호"),
    size: int = Query(20, ge=1, le=100, description="페이지 크기"),
    db: Session = Depends(get_db)
):
    """의류 아이템 검색"""
    try:
        skip = (page - 1) * size
        items, total = clothing_crud.search_clothing_items(
            db=db, 
            query=q, 
            skip=skip, 
            limit=size
        )
        
        total_pages = math.ceil(total / size)
        
        return ClothingItemsListResponse(
            items=items,
            total=total,
            page=page,
            size=size,
            total_pages=total_pages
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"검색 중 오류가 발생했습니다: {str(e)}")

@router.get("/{product_id}", response_model=ClothingItemResponse)
async def get_clothing_item(
    product_id: int,
    db: Session = Depends(get_db)
):
    """특정 의류 아이템 조회"""
    try:
        item = clothing_crud.get_clothing_item_by_id(db=db, product_id=product_id)
        if not item:
            raise HTTPException(status_code=404, detail="상품을 찾을 수 없습니다")
        return item
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"상품 조회 중 오류가 발생했습니다: {str(e)}")

File: frontend/src/api/clothing_items.js
/**
 * ===================================================================
 * 의류 아이템 관리 API
 * ===================================================================
 *
 * 주요 기능:
 * - 의류 아이템 목록 조회 (페이지네이션, 정렬, 필터링)
 * - 인기/최신 상품 조회
 * - 의류 검색 기능
 * - 카테고리별/성별 상품 분류
 * - 브랜드별 상품 필터링
 * - 고급 브라우징 (다중 필터 조합)
 *
 * 기술 스택:
 * - HTTP Client: Axios
 * - 인증: Bearer Token (localStorage)
 *
 * 사용 예시:
 * - 메인 페이지 상품 진열
 * - 가상 피팅을 위한 의류 선택
 * - 카테고리별 상품 브라우징
 * - 검색 결과 페이지
 * - 상품 상세 페이지
 * - 좋아요 기능과 연동
 *
 * 지원 필터:
 * - 메인 카테고리 (상의, 하의, 아우터 등)
 * - 서브 카테고리 (티셔츠, 청바지 등)
 * - 성별 (남성, 여성, 유니섹스)
 * - 브랜드
 * - 정렬 (인기순, 최신순, 가격순)
 * ===================================================================
 */

import axios from "axios"

const API_BASE_URL = process.env.REACT_APP_API_URL || ""

// Axios 인스턴스 생성
const clothingAPI = axios.create({
  baseURL: `${API_BASE_URL}/api/clothing`,
  timeout: 10000,
})

// 요청 인터셉터 (인증 토큰 추가)
clothingAPI.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem("accessToken")
    if (token) {
      config.headers.Authorization = `Bearer ${token}`
    }
    return config
  },
  (error) => {
    return Promise.reject(error)
  },
)

// 응답 인터셉터 (에러 처리)
clothingAPI.interceptors.response.use(
  (response) => response,
  (error) => {
    console.error("API 오류:", error)
    return Promise.reject(error)
  },
)

/**
 * 의류 아이템 목록 조회
 */
export const getClothingItems = async (params = {}) => {
  try {
    const { page = 1, size = 20, sort_by = "likes", order = "desc", category = null, gender = null } = params

    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sort_by,
      order,
    })

    if (category) queryParams.append("category", category)
    if (gender) queryParams.append("gender", gender)

    const response = await clothingAPI.get(`/?${queryParams}`)
    return response.data
  } catch (error) {
    console.error("의류 아이템 목록 조회 실패:", error)
    throw error
  }
}

/**
 * 인기 상품 조회
 */
export const getPopularItems = async (limit = 6) => {
  try {
    const response = await clothingAPI.get(`/popular?limit=${limit}`)
    return response.data
  } catch (error) {
    console.error("인기 상품 조회 실패:", error)
    throw error
  }
}

/**
 * 최신 상품 조회
 */
export const getLatestItems = async (limit = 6) => {
  try {
    const response = await clothingAPI.get(`/latest?limit=${limit}`)
    return response.data
  } catch (error) {
    console.error("최신 상품 조회 실패:", error)
    throw error
  }
}

/**
 * 상품 검색
 */
export const searchClothingItems = async (query, page = 1, size = 20) => {
  try {
    const response = await clothingAPI.get(`/search?q=${encodeURIComponent(query)}&page=${page}&size=${size}`)
    return response.data
  } catch (error) {
    console.error("상품 검색 실패:", error)
    throw error
  }
}

/**
 * 카테고리 정보 조회
 */
export const getCategories = async () => {
  try {
    const response = await clothingAPI.get("/categories")
    return response.data
  } catch (error) {
    console.error("카테고리 조회 실패:", error)
    throw error
  }
}

/**
 * 특정 상품 조회
 */
export const getClothingItem = async (productId) => {
  try {
    const response = await clothingAPI.get(`/${productId}`)
    return response.data
  } catch (error) {
    console.error("상품 조회 실패:", error)
    throw error
  }
}

/**
 * 카테고리별 상품 조회 (편의 함수)
 */
export const getClothingItemsByCategory = async (category, params = {}) => {
  return getClothingItems({ ...params, category })
}

/**
 * 성별별 상품 조회 (편의 함수)
 */
export const getClothingItemsByGender = async (gender, params = {}) => {
  return getClothingItems({ ...params, gender })
}

/**
 * 필터링과 검색이 가능한 의류 아이템 브라우징
 */
export const browseClothingItems = async (params = {}) => {
  try {
    const {
      page = 1,
      size = 20,
      sort_by = "likes",
      order = "desc",
      main_category = null,
      sub_category = null,
      gender = null,
      brand = null,
      search = null,
    } = params

    const queryParams = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sort_by,
      order,
    })

    if (main_category) queryParams.append("main_category", main_category)
    if (sub_category) queryParams.append("sub_category", sub_category)
    if (gender) queryParams.append("gender", gender)
    if (brand) queryParams.append("brand", brand)
    if (search) queryParams.append("search", search)

    const response = await clothingAPI.get(`/browse?${queryParams}`)
    return response.data
  } catch (error) {
    console.error("상품 브라우징 실패:", error)
    throw error
  }
}